tiledb/sm/filter/filter.cc

```diff
@@ -106,28 +117,38 @@ Status Filter::set_option(FilterOption option, const void* value) {
   return set_option_impl(option, value);
 }
 
-Status Filter::deserialize(ConstBuffer* buff, Filter** filter) {
+std::tuple<Status, std::optional<std::shared_ptr<Filter>>> Filter::deserialize(
+    ConstBuffer* buff) {
+  Status st;
   uint8_t type;
-  RETURN_NOT_OK(buff->read(&type, sizeof(uint8_t)));
+  st = buff->read(&type, sizeof(uint8_t));
+  if (!st.ok()) {
+    return {st, std::nullopt};
+  }
   uint32_t filter_metadata_len;
-  RETURN_NOT_OK(buff->read(&filter_metadata_len, sizeof(uint32_t)));
+  st = buff->read(&filter_metadata_len, sizeof(uint32_t));
+  if (!st.ok()) {
+    return {st, std::nullopt};
+  }
 
-  auto* f = create(static_cast<FilterType>(type));
-  if (f == nullptr)
-    return LOG_STATUS(Status::FilterError("Deserialization error."));
+  auto&& [st_filter, filter]{create(static_cast<FilterType>(type))};
+  if (!st_filter.ok()) {
+    return {st_filter, std::nullopt};
+  }
 
   auto offset = buff->offset();
-  RETURN_NOT_OK_ELSE(f->deserialize_impl(buff), tdb_delete(f));
+  st = filter.value()->deserialize_impl(buff);
+  if (!st.ok()) {
+    return {st_filter, std::nullopt};
+  }
 
   if (buff->offset() - offset != filter_metadata_len) {
-    tdb_delete(f);
-    return LOG_STATUS(Status::FilterError(
-        "Deserialization error; unexpected metadata length"));
+    return {Status::FilterError(
+                "Deserialization error; unexpected metadata length"),
+            std::nullopt};
```

Comment:
To future-proof this function, we don't want to be checking the length after the fact by checking buffer offsets after reading. Instead, we want to check in advance that the buffer contains at least as many characters as would be expected.

"In advance" here doesn't mean a single check. Before we read the type in, we know that the buffer needs to be at least as long as the integer that holds the length, so we can check that in advance. After we read the type, we know how many characters the particular type will need, so we also check that, in a second check, in advance of reading.

Reply:
changed to in-advance check